utils/merge_test.go

```diff
@@ -0,0 +1,829 @@
+		merged, err := mergeStringIntMap(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge maps where patch is empty", func(t *testing.T) {
+		m1 := map[string]int{"this": 1, "is": 3, "a second map": 3, "another key": 4}
+		m2 := make(map[string]int)
+		expected := map[string]int{"this": 1, "is": 3, "a second map": 3, "another key": 4}
+		merged, err := mergeStringIntMap(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+}
+
+func TestMergeWithSlices(t *testing.T) {
+	t.Run("merge slices where patch is longer", func(t *testing.T) {
+		m1 := []string{"this", "will", "be", "overwritten"}
+		m2 := []string{"this one", "will", "replace the other", "one", "and", "is", "longer"}
+		expected := []string{"this one", "will", "replace the other", "one", "and", "is", "longer"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge slices where base is longer", func(t *testing.T) {
+		m1 := []string{"this", "will", "be", "overwritten", "but", "not", "this"}
+		m2 := []string{"this one", "will", "replace the other", "one"}
+		expected := []string{"this one", "will", "replace the other", "one", "but", "not", "this"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge slices where base is empty slice", func(t *testing.T) {
+		m1 := []string{}
+		m2 := []string{"this one", "will", "replace the other", "one"}
+		expected := []string{"this one", "will", "replace the other", "one"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge slices where base is nil", func(t *testing.T) {
+		var m1 []string
+		m2 := []string{"this one", "will", "replace the other", "one"}
+		expected := []string{"this one", "will", "replace the other", "one"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge slices where patch is empty struct", func(t *testing.T) {
+		m1 := []string{"this", "will", "be", "overwritten", "but", "not", "this"}
+		m2 := []string{}
+		expected := []string{"this", "will", "be", "overwritten", "but", "not", "this"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+
+	t.Run("merge slices where patch is nil", func(t *testing.T) {
+		m1 := []string{"this", "will", "be", "overwritten", "but", "not", "this"}
+		var m2 []string
+		expected := []string{"this", "will", "be", "overwritten", "but", "not", "this"}
+		merged, err := mergeStringSlices(m1, m2)
+		require.NoError(t, err)
+
+		assert.Equal(t, expected, merged)
+	})
+}
+
+func TestMergeWithStructs(t *testing.T) {
+	t.Run("merge identical structs - simple", func(t *testing.T) {
+		t1 := simple{42, 42.2, newFloat64(932.2), newInt(45), newBool(true),
```

Comment:
nit: I might recommend explicitly naming the fields here, as it's hard to know which are exported looking at the initialization alone.

Reply:
I was thinking of just keeping these at the bottom of the test file because they are kind of 'comprehensive' -- but they are definitely too much. So I made a few more specialized tests. Let me know if there's a case that I'm missing.